include/dmlc/endian.h

```diff
@@ -8,7 +8,7 @@
 
 #include "./base.h"
 
-#if defined(__APPLE__) || defined(_WIN32)
+#if defined(__APPLE__) || defined(__FreeBSD__) || defined(_WIN32)
```

Comment:
FreeBSD is not necessarily little endian. See https://lists.freebsd.org/pipermail/freebsd-questions/2008-February/167918.html.
To determine endianness of FreeBSD, we can include `<sys/endian.h>`

Reply:
FYI: `<sys/endian.h>` is available on all BSDs, according to this doc: http://man7.org/linux/man-pages/man3/endian.3.html

Reply:
@mallniya In the future, I will add FreeBSD to CI tests to ensure that dmlc-core compiles on FreeBSD. For now, let me update this PR to include better endian detection logic, as found in https://www.boost.org/doc/libs/1_54_0/boost/detail/endian.hpp

Reply:
@mallniya Can you try compiling with my changes?

Reply:
@hcho3 yes, compilation succeed.

Reply:
Cool!